### ISSUE TITLE
List.groupBy doesn't preserve order

### ISSUE THREAD

**Issue by davidtme** [fable-compiler/Fable#1738]
### Description

Using the `groupBy` seem to change the order of the list which doesn't match what happens in .net

### Repro code

```
[ 0; 1; 2; 3 ]
|> List.groupBy(id)
|> List.iteri(fun index (_, item) -> printfn "%d = %d" index (item |> List.head))
```

### Expected  (via f# interactive)
```
0 = 0
1 = 1
2 = 2
3 = 3
```

### Actual (via repl)
```
0 = 1
1 = 3
2 = 0
3 = 2
```

### Related information

* Fable version: fable-compiler 2.1.11

**Comment by alfonsogarciacaro** [comment]
We're using a dictionary for the grouping. However, because the dictionary needs to accept a custom equality comparer, we're implementing it with F# Map which doesn't honor the key order. We need to find a way to retain this information, probably saving the indices somewhere.

**Comment by alfonsogarciacaro** [comment]
Do you want to give it a go? I think `Array.groupBy` has the same problem. https://github.com/fable-compiler/Fable/blob/16ab2eed8b213836a9fd1e3253c1c4d633061e74/src/fable-library/List.fs#L445-L452

**Comment by ncave** [comment]
Fixed by #1754.
